internal/providers/vault/provider.go

```diff
@@ -39,53 +42,71 @@ func (p *Provider) GetName() string {
 	return p.Name
 }
 
-// VaultDefaultField is the default value returned by the provider from the
-// hash returned by the Vault.
-const VaultDefaultField = "value"
+// DefaultField is the default field name the provider expects to find the secret value.
+const DefaultField = "value"
 
-// parseVaultID returns the secret id and field name.
-func parseVaultID(id string) (string, string) {
+// parseVaultID returns the path to the secret (object) and (normalized) field/property path to the secret value.
+func parseVaultID(id string) (string, []string) {
 	tokens := strings.SplitN(id, "#", 2)
 	switch len(tokens) {
 	case 1:
-		return tokens[0], VaultDefaultField
+		return tokens[0], []string{DefaultField}
 	default:
-		return tokens[0], tokens[1]
+		return tokens[0], strings.Split(tokens[1], ".")
 	}
 }
 
-// GetValue obtains a value by id. Any secret which is stored in the vault is recognized.
-// The datatype returned by Vault is map[string]interface{}. Therefore this provider needs
-// to know which field to return from the map. By default, it returns the 'value'.
-// An alternative field can be obtained by appending '#fieldName' to the id argument.
-func (p *Provider) GetValue(id string) (value []byte, err error) {
-	id, fieldName := parseVaultID(id)
-
-	var secret *vault.Secret
-	if secret, err = p.Client.Logical().Read(id); err != nil {
-		return
-	}
-	// secret can be nil if it's not found
-	if secret == nil {
-		err = fmt.Errorf("HashiCorp Vault provider could not find a secret called '%s'", id)
-		return
+// parseSecret returns value navigated by given fields on secret object.
+// Note that a secret returned from Vault is effectively a JSON object.
+func parseSecret(secret *vault.Secret, path string, fields []string) ([]byte, error) {
+	// navigate to secret value
+	var obj, ok = secret.Data, false
+	for _, field := range fields[:len(fields)-1] {
+		obj, ok = obj[field].(map[string]interface{})
```

Comment:
I'm a bit confused why this value is declared outside of the `for` block if the check of it's value is only done immediately after this line here.

Reply:
I think this loop also needs a bit clearer logic about what it does either with more obvious code (preferred) or some really good comments

Reply:
I rewrote / refactored. Effectively it takes an object and a path traversing through the object structure. Suppose obj = { "foo": { "bar": "qux" } } and fields = "foo.bar", then returned value is "qux". It navigates from outer to inner object and last field's value.

Reply:
Oh for sure - I wasn't confused about what you were trying to do. My comment was more to ensure others can easily understand what is done in the flow if they are reading this for the first time.